plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -135,7 +135,9 @@ public class DefaultJdbcMetadata
     private final boolean precalculateStatisticsForPushdown;
     private final Set<JdbcQueryEventListener> jdbcQueryEventListeners;
 
-    private final AtomicReference<Runnable> rollbackAction = new AtomicReference<>();
+    private final AtomicReference<Runnable> insertRollbackAction = new AtomicReference<>();
```

Comment:
I don't think this will work. Consider multiple statements running in parallel. Changing a state for one query may affect the other query that is running.

Reply:
For the different query will has its own metadata. Do you mean the `query` is the part of the `MERGE`, i.,e its split into the `insert`, `update` and `delete` part?